lib/pbench/test/unit/server/query_apis/test_datasets_list.py

```diff
@@ -191,6 +218,31 @@ def test_query(
         assert data[0]["end"] == "2020-04-29T13:30:04.918704"
         assert data[0]["id"] == "12fb1e952fd826727810868c9327254f"
 
+        # Test without authorization header
+        # All tests should pass except when the non-empty user field is present in the json
+        if user != "drb":
+            response = query_api(
+                "/datasets/list",
+                "/_search?ignore_unavailable=true",
+                json,
+                index,
+                HTTPStatus.OK,
+                authorization=False,
+                status=HTTPStatus.OK,
+                json=response_payload,
+            )
+            res_json = response.json
+            assert len(res_json.keys()) == 1
+            data = res_json["dhcp31-187.example.com"]
+            assert isinstance(data, list)
+            assert len(data) == 1
+
+        else:
+            response = client.post(
+                f"{server_config.rest_uri}/datasets/list", json=json,
+            )
+            assert response.status_code == HTTPStatus.FORBIDDEN
```

Comment:
okay I refactored a little bit so `authentication` can be received as a parameterized argument but I still need `if-else` condition because pytest gonna create all the combinations of `user` and `authentication` field e.g. `(user: "", authentication: True), (user: "", authentication: False), (user: None, authentication: True), ...` and all these tests should be `HTTPStatus.OK` except `(user: "drb", authentication: False)` this should be `401`

Reply:
You should be able to add a parameter for the expected result to the parameterization (to the outer one, if the expected result is independent of the inner one, or to the inner one possibly incorporating the value of the outer one if it's available in that context, or, failing those, you can generate the expected result in the function body, based on the two parameterization values).  It's perfectly reasonable to have an `if-else` construct for determining the expected result; what I'm objecting to is having a conditional affecting how the CUT is invoked (`client.post()` vs. `query_api()`) -- the parameterization should take care of that.

Reply:
We can certainly add expected result as a parameterization but it doesn't add much value, because in this test the only time the expected result would be different than the `OK` is when we encounter this scenario `(user: "drb", authentication: False)`. Also the `query_api` internally does `client.post()` so they are not different per se. The only difference is that `query_api` builds a mock response with an expected index and other user provided parameter such as response payload, status, etc on top of the POST. Even if I use the `query_api` for `(user: "drb", authentication: False)` I have to use different set of parameters such as I can't pass the response payload (as there wont be any payload in the response since its 403) and also doing `rsp.add(responses.POST, es_url, **kwargs)` in `query_api` results in an error because in our definition of `post` (lib/pbench/server/api/resources/query_apis/__init__.py) we abort early if we encounter 403 and so wont proceed to execute `python.request`.

Reply:
> they are not different per se.

It's the "per se" part that I'm stuck on.  The fact that they are taking distinct paths to generate the result poses problems:  as the code evolves, it's possible for cracks to open between the two paths where things go untested, and it's hard for the reader to discern what coverage the two paths generate.  It's better if we can avoid multiple paths on the driver side and have them only in the part which analyzes the returns.

Regarding parameterizing the expected value, it's a toss up:  having the expected values laid out in a tabular format has nice power, but, if there's only one scenario out of the dozen which produces a different result, then perhaps it would be better to special-case the generation of that value.

With regard to the rest of it, I think it boils down to having the test code express its expectations appropriately.  That is, the expectations are sort of a bundle:  we expect a certain status return, we expect a certain message and/or other returned payload, and we expect certain behavior with respect to requests made by the CUT.  Thus, if we set up a response to the server's request, and then we get a 403 (which we expect), we should also expect that the response that we queued will be unconsumed -- isn't there a way to test/assert that the response is unconsumed (and thereby consume/remove it) without resulting in a test failure?

Reply:
> isn't there a way to test/assert that the response is unconsumed (and thereby consume/remove it) without resulting in a test failure?

`query_api()` uses the `responses.RequestsMock()` as a context manager, which is nice in terms of encapsulation, but inconvenient in terms of implementing assertions in the calling test code.  The object could be instead be implemented as a persistent object or even as [a PyTest fixture](https://pypi.org/project/responses/#responses-as-a-pytest-fixture), which would allow the test to check whether the responses have been consumed or not.  (Also, there are `reset()` and `remove()` methods which could be used to clean up after the test, and there is an option on the c'tor to suppress the unused requests assertion...so there are various options.)

Reply:
Okay for some reason github just didn't load this comment (or my browser was loading the old page from cache). Anyway I added a conditional check to see if the expected status is `FORBIDDEN` or `UNAUTHORIZED` since these are the 2 conditions when we would abort early. So if the expected status is not one of these 2, we would add the response body to the mock otherwise we skip it 
```
if expected_status not in [HTTPStatus.FORBIDDEN, HTTPStatus.UNAUTHORIZED]:
    rsp.add(responses.POST, es_url, **kwargs)
```